odoo/addons/base/models/qweb.py

```diff
@@ -397,397 +243,325 @@ def get_template(self, template, options):
         """
-        return ast.parse(dedent("""
-            from collections import OrderedDict
-            from html import unescape # not sure whether this one or werkzeug's is better
-            from markupsafe import escape, Markup
-            from odoo.tools.pycompat import to_text
-            from odoo.addons.base.models.qweb import foreach_iterator
-            """))
-
-    def _create_def(self, options, body, prefix='fn', lineno=None):
-        """ Generate (and globally store) a rendering function definition AST
-        and return its name. The function takes parameters ``self``, ``append``,
-        ``values``, ``options``, and ``log``. If ``body`` is empty, the function
-        simply returns ``None``.
+        return values
+
+    def _prepare_globals(self, globals_dict, options):
+        """ Prepare the global context that will sent to eval the qweb generated
+        code.
+
+        :param values: template values to be used for rendering
+        :param options: frozen dict of compilation parameters.
         """
-        #assert body, "To create a compiled function 'body' ast list can't be empty"
-
-        name = self._make_name(prefix)
-
-        # def $name(self, append, values, options, log)
-        fn = ast.FunctionDef(
-            name=name,
-            args=ast.arguments(args=[
-                ast.arg(arg='self', annotation=None),
-                ast.arg(arg='append', annotation=None),
-                ast.arg(arg='values', annotation=None),
-                ast.arg(arg='options', annotation=None),
-                ast.arg(arg='log', annotation=None),
-            ], defaults=[], vararg=None, kwarg=None, posonlyargs=[], kwonlyargs=[], kw_defaults=[]),
-            body=body or [ast.Return()],
-            decorator_list=[])
-        if lineno is not None:
-            fn.lineno = lineno
-
-        options['ast_calls'].append(fn)
-
-        return name
-
-    def _call_def(self, name, append='append', values='values'):
-        # $name(self, append, values, options, log)
-        return ast.Call(
-            func=ast.Name(id=name, ctx=ast.Load()),
-            args=[
-                ast.Name(id='self', ctx=ast.Load()),
-                ast.Name(id=append, ctx=ast.Load()) if isinstance(append, str) else append,
-                ast.Name(id=values, ctx=ast.Load()),
-                ast.Name(id='options', ctx=ast.Load()),
-                ast.Name(id='log', ctx=ast.Load()),
-            ],
-            keywords=[], starargs=None, kwargs=None
-        )
+        globals_dict['Sized'] = Sized
+        globals_dict['Mapping'] = Mapping
+        globals_dict['Markup'] = Markup
+        globals_dict['escape'] = escape
+        globals_dict['type'] = type
+        globals_dict['format'] = self._format
+        globals_dict['compile_options'] = options
+        globals_dict.update(self._available_objects)
+        if options.get('dev_mode'):
+            globals_dict['__import__'] = __import__
```

Comment:
this must be replaced by a specific function that can only run a debugger, e.g. something like the following `_debug_trace()` function, which can then be used by code generated by `t-debug`:

```py
# same list as for `--dev-mode` ...
SUPPORTED_DEBUGGERS = ['pdb', 'ipbd', 'pudb', 'wdb']

def _debug_trace(debugger):
    if debugger in SUPPORTED_DEBUGGERS:
        __import__(debugger).set_trace()
    else:
        raise ValueError(f"unsupported t-debug debugger: {debugger}")
```